app.go

```diff
@@ -1,248 +1,226 @@
-               _, err := listener.Unsubscribe(topicID)
-               if err != nil {
-                   log.Warn().Msg("Failed to unsubscribe")
-               } else {
-                   log.Debug().Msg("Event listener successfully terminated")
-               }
-               return
-           case eventMessage, ok := <-events:
-               if !ok {
-                   log.Info().Msg("Failed to read events")
-                   break
-               }
-               if len(eventMessage.Events) == 0 {
-                   log.Debug().Msg("Gotta event message with no events")
-                   break
-               }
-               log.Debug().Msgf("Processing %+v events", len(eventMessage.Events))
-               for _, event := range eventMessage.Events {
-                   go app.processEvent(event)
-               }
-               offset = eventMessage.Events[len(eventMessage.Events) - 1].Offset + 1
-               writeOffset(offsetPath, offset)
-           }
-        }
-    }(parentContext)
+func (app *App) RunEventProcessor(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case eventMessage, ok := <-app.EventMessages:
+			if !ok {
+				log.Warn().Msg("Failed to read events")
+				break
+			}
+			if len(eventMessage.Events) == 0 {
+				log.Debug().Msg("Gotta event message with no events")
+				break
+			}
+			log.Debug().Msgf("Processing %+v events", len(eventMessage.Events))
+			for _, event := range eventMessage.Events {
+				go app.processEvent(event)
+			}
+			offset := eventMessage.Events[len(eventMessage.Events)-1].Offset + 1
+			if err := writeOffset(app.OffsetHandler, offset); err != nil {
+				log.Error().Msgf("Failed to write offset, reason: %s", err.Error())
+			}
+		}
+	}
 }
 
-func (app *App) Run(addr string) {
-    parentContext, cancel := context.WithCancel(context.Background())
-    log.Debug().Msg("starting http server")
-    go func() {
-        log.Error().Msg(graceful.ListenAndServe(addr, app.Router).Error())
-    }()
-    log.Debug().Msg("stating event listener")
-    var wg sync.WaitGroup
-    wg.Add(1)
-    app.RunEventListener(parentContext, &wg)
-
-    // Handle signals
-    done := make(chan os.Signal, 1)
-    signal.Notify(done, syscall.SIGINT, syscall.SIGTERM)
-    log.Debug().Msg("Waiting for signal")
-    <-done
-    log.Info().Msg("Terminating service")
-    cancel()
-    wg.Wait()
-    log.Info().Msg("Service successfully terminated")
+func (app *App) Run(addr string) error {
+	go func() {
+		log.Debug().Msg("starting http server")
+		log.Panic().Msg(graceful.ListenAndServe(addr, app.GetRouter()).Error())
+	}()
+
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+	defer func() {
+		log.Info().Msg("Terminating service")
+		cancel()
+		wg.Wait()
+		log.Info().Msg("Service successfully terminated")
+	}()
+
+	log.Debug().Msg("starting event listener")
```

Comment:
mb info ?

Reply:
I don't think that's really important
all I want to see whether the service started successfully and I log that with info level

Reply:
ok